### DISCUSSION TITLE
Multiple simultaneous queries within an async session

### DISCUSSION THREAD

**Discussion by SantjagoCorkez** [sqlalchemy/sqlalchemy#6965] [Usage Questions] [answered]
SQLAlchemy version: 1.4.19

It's unclear from the docs if it's possible to run parallel queries against a PostgreSQL using all the underlying Pool capacity to speed-up.

For example, the code like this:
```python
engine = create_async_engine(url, future=True)
session_maker = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession, future=True)
async with session_maker() as session, session.begin():
  queries = split_into_chunks(delete(my_table).where, large_values_set)
  if statements:
    await asyncio.gather(*(session.execute(s) for s in statements))
```

executes within a single connection according to PostgreSQL logs. That means queries are run one-by-one, though it's possible to speed-up execution utilizing the whole underlying Pool's capacity and spawn as many connections as possible and run all the queries against those connections cycling them if the queries amount is overwhelming the pool capacity. That worked well with plain asyncpg pool. Of course, as usual, avoiding deadlocks is the user's responsibility so it's out of the scope of this discussion.

**Comment by zzzeek** [comment]
a Session is per-connection so you would need to use a distinct Session for each awaitable passed to gather().  the results from each can be merged into one session using [loading.merge_frozen_result](https://docs.sqlalchemy.org/en/14/orm/internals.html?highlight=merge_frozen_result#sqlalchemy.orm.merge_frozen_result).

**Reply by SantjagoCorkez** [reply]
So, do I understand correctly that the code should look something like the following

```python
async def parallel_run(query):
  async with session_maker() as session, session.begin():
    await session.execute(query)

engine = create_async_engine(url, future=True)
session_maker = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession, future=True)
queries = split_into_chunks(delete(my_table).where, large_values_set)
if queries:
  await asyncio.gather(*(parallel_run(q) for q in queries))
```

and that will guarantee the maximum pool capacity load along with correct spawned sessions dismissing?

Also, could it be then a request to update the docs to reflect your clarification (if there's not already such a statement which I could miss, of course)?

**Reply by CaselIT** [reply]
> Also, could it be then a request to update the docs to reflect your clarification (if there's not already such a statement which I could miss, of course)?

sure, feel free of opening a PR for the docs. I think the docs mention this, but I could not find it at the moment

**Reply by zzzeek** [reply]
you know what, there is a better way to do this where you can use just the one Session, because that freeze/merge step I mention is going to be very inefficient.   I consider it to be total heresy however asyncio people are going to keep asking for this, let me just work up the example.

**Reply by CaselIT** [reply]
> because that freeze/merge step I mention is going to be very inefficient

Surely `itertool.chains` could be used as an alternative, to avoid having to merge the result set in an single one if all it's needed is a list of things.

**Reply by zzzeek** [reply]
@CaselIT the issue is that all those objects will be detached from the current session, which may or may not be a problem. 

my idea to use just one session is not going to work right now without changes made to the Session and in any case it's probably a good idea not to encourage concurrent writes to a single session.

regardless, I think people should at least use autocommit connections and this and such so let me write a recipe that is the most efficient way to do it right now.

**Reply by CaselIT** [reply]
> @CaselIT the issue is that all those objects will be detached from the current session, which may or may not be a problem.

yes, that would be a problem if things are orm objects

**Comment by zzzeek** [comment, marked as answer]
I've worked up the best case form of this for now in the review at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3042 .    If you can work with this program to see that it does everything needed, we can merge this as a new example on the site.

**Reply by SantjagoCorkez** [reply]
Did not test it yet but the code looks as what I myself needed opening this discussion. I'll mark your recipe as the answer.

**Reply by SantjagoCorkez** [reply]
Also, I understand the contras you described in the header of the example. Though there are or could be many situations where this approach will definitely do.

For example:
You have a request for some aggregates (distinct for each value of a column's values' subset). It sometimes will be faster to run 10 parallel queries within PostgreSQL each one for each distinct grouping column (which is then a WHERE clause filter also: `SELECT some_agg(column1) FROM some_table WHERE grouping_column = v1`, and the same for `v2`, `v3`, ...) rather then trying to query `SELECT some_agg(column1), grouping_column FROM some_table WHERE grouping_column IN (v1, v2, v3, v4, ...)` since sometimes PostgreSQL: gets messed with indexes usage, decides to first get the rows into temporary table and aggregate them only after selecting and so on. Querying small datasets could improve the overall time of execution, by avoiding unneeded hard-drive offloaded temporary tables writes.

Of course, the above does not cancel your description and a user must decide every time one wants to split-execute queries in parallel as the example's description states.

**Reply by CaselIT** [reply]
That case it's probably faster if using a single query with the various selects in an union_all clause

**Reply by zzzeek** [reply]
the "contras" I mention are about as reserved as I could get in how I expressed them :)    the pattern you have there I feel is one that people really assume is always superior, as they are looking at "query the database" the same way they look at it as "query a web service".   But especially for local database connections, most of the time it's not appropriate, given the loss of transactional semantics coupled with making new DB connections is not optimized the way HTTP requests are; database connections are in every way the opposite of HTTP.

ideally, when you use this pattern, you'd measure it against the "normal" way and see that it does in fact improve performance.    There's going to be a threshold where it does and where it doesnt.

**Reply by SantjagoCorkez** [reply]
> That case it's probably faster if using a single query with the various selects in an union_all clause

True. Until we grow a monster of our union_all parameter. As you see in my starting question, I've already stuck with the values sets too large for PostgreSQL itself (in default build config) so I need to split my queries into chunks. The same for `UNION ALL` and the larger the text of the query is the slower the initialization of the backend is even before starting to plan the query. I once have made my local dev instance of PostgreSQL stuck with parsing something like 100MiB query (by mistake of course). It didn't even start to plan the query after half an hour.

Anyway, thanks to you all for the explainations and advices.

**Reply by CaselIT** [reply]
If the query string is that big (you mentioned that you overflow the bound parameters, that iirc are 32k in pg) it may make sense to use a temp table and then select from there in the main query?

**Reply by SantjagoCorkez** [reply]
Sure, that's what I ended up with. But PostgreSQL slows down not on parameters list only, but on tokens themselves since it must parse them all and validate the whole query. The longer the query string is, the more complex tree it builds while is being parsed and validated.

**Reply by CaselIT** [reply]
Sadly I guess that's common to all sql based dbs, since sql is text based
